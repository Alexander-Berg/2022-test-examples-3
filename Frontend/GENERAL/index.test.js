const getHashes = require('./index');

describe('getHashes()', () => {
    it('should generate hashes', () => {
        const html = `
<!DOCTYPE html>
<html lang="ru">
    <head>
        <meta charset="utf-8" />
        <meta
            name="viewport"
            content="width=device-width,initial-scale=1,shrink-to-fit=no,user-scalable=no,viewport-fit=cover"
        />
        <meta name="theme-color" content="#000000" />
        <meta name="robots" content="noindex,nofollow" />
        <link rel="icon" href="/favicon.ico" />
        <link href="/static/css/main.f530a6b6.chunk.css" rel="stylesheet" />
        <style>
            body { visibility: hidden; }
        </style>
    </head>
    <body>
        <noscript>You need to enable JavaScript to run this app.</noscript>
        <div id="root"></div>
        <script>
            console.log('important script!');
        </script>
        <script src="/static/js/main.476b6822.chunk.js"></script>
    </body>
</html>
    `;

        expect(getHashes(html)).toEqual({
            scripts: [
                "'sha256-8GrLtCdMH+SStlLKH6vDvj0ikVByaagWhdasX+hKPcc='",
            ],
            styles: [
                "'sha256-l0OAVvAnhhGrFDApKn807b3Feu8oLP9Q7AlAJ2SHy7c='"
            ],
        });
    });

    it('should works with empty scripts', () => {
        const html = `
    <script></script>
    <script type="text/javascript"></script>
    <style></style>
    <style type="text/css"></style>
    `;

        expect(getHashes(html)).toEqual({
            scripts: [
                "'sha256-47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU='",
            ],
            styles: [
                "'sha256-47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU='"
            ],
        });
    });

    it('should ignore external scripts', () => {
        const html = `
    <script src="index.js"></script>
    <script type="text/javascript" src="index.js"></script>
    `;

        expect(getHashes(html)).toEqual({
            scripts: [],
            styles: [],
        });
    });

    it('should works with empty page', () => {
        const html = '';

        expect(getHashes(html)).toEqual({
            scripts: [],
            styles: [],
        });
    });
});
