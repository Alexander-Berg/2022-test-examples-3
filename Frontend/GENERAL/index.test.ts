import { ABTSplitter, ABTInfo, ABTRawConfig } from '.';

function makeHeadersByInfo(info: Partial<ABTInfo>): Record<string, string> {
    if (!info.configs) {
        throw new Error('Param `configs` is required');
    }

    return {
        'x-yandex-expconfigversion': String(info.version || '169'),
        'x-yandex-randomuid': String(info.randomUid || ''),
        'x-yandex-expflags': (info.configs as ABTRawConfig[]).map(raw => Buffer.from(JSON.stringify(raw), 'ascii').toString('base64')).join(','),
        'x-yandex-expboxes': info.boxes || '',
        'x-yandex-expboxes-crypted': String(info.experiments || ''),
    };
}

describe('Splitter', () => {
    it('should use the right context and correctly parse flags', () => {
        const splitter = new ABTSplitter({ service: 'svc', handler: 'SVC' });
        const configs = [{
            HANDLER: 'SVC',
            CONTEXT: {
                SVC: { flags: ['expflag=goodvalue', 'bool=true', 'valid'], testid: ['1'] },
                NOTSVC: { flags: ['expflag=badvalue', 'invalid'], testid: ['2'] },
            },
            // 'desktop && ((device.BrowserName ne "MSIE" || device.BrowserVersion gt "10"))'
            CONDITION: 'UNUSED',
            RESTRICTIONS: [{ services: 'svc' }],
        }];
        const experiments = 'SOME THRING';

        const info = splitter.parseInfo(makeHeadersByInfo({ configs, experiments }));

        expect(info).toMatchObject({ experiments });
        expect(info.configs).toEqual(configs);
        expect(info.flags).toEqual({ expflag: 'goodvalue', bool: true, valid: true });
    });
});
