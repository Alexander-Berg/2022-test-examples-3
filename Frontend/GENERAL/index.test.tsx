import {
    findIndexFromHead,
    findIndexFromTail,
    setChoicesSubset,
    delChoicesSubset,
    addChoicesSubset,
    hasChoicesSubset,
    findPrevPicked,
    findNextPicked,
    eventHandler,
    scrollToElem,
} from '.';

describe('ToolsSuggest.utils', () => {
    describe('scrollToElem()', () => {
        it('Should not fail on falsey elem', () => {
            expect(() => {
                scrollToElem(null);
            }).not.toThrow();
        });

        it('Should scroll to elem', () => {
            jest.useFakeTimers();
            const elem = document.createElement('div');

            elem.scrollIntoView = jest.fn();

            scrollToElem(elem);
            jest.runAllTimers();

            expect(elem.scrollIntoView).toHaveBeenCalledWith({
                behavior: 'auto',
                block: 'nearest',
                inline: 'nearest',
            });
        });
    });

    describe('eventHandler()', () => {
        it('Should not fail', () => {
            expect(() => {
                eventHandler();
            }).not.toThrow();
        });
    });

    describe('findIndexFromHead()', () => {
        it('Should find first matched choice', () => {
            expect(findIndexFromHead([{ id: '1' }, { id: '2' }, { id: '1' }, { id: '2' }], { id: '1' })).toBe(0);
        });

        it('Should support undefined choices', () => {
            expect(findIndexFromHead(undefined, { id: '1' })).toBe(-1);
        });
    });

    describe('findIndexFromTail()', () => {
        it('Should find last matched choice', () => {
            expect(findIndexFromTail([{ id: '1' }, { id: '2' }, { id: '1' }, { id: '2' }], { id: '1' })).toBe(2);
        });

        it('Should support undefined choices', () => {
            expect(findIndexFromTail(undefined, { id: '1' })).toBe(-1);
        });
    });

    describe('setChoicesSubset()', () => {
        it('Should replace choices', () => {
            expect(setChoicesSubset([{ id: '1' }, { id: '2' }], [{ id: '1' }], Infinity)).toEqual([{ id: '1' }]);
        });

        it('Should crop choices to maxChoices from head', () => {
            expect(setChoicesSubset([], [{ id: '1' }, { id: '2' }], 1)).toEqual([{ id: '2' }]);
        });

        it('Should support corner cases for maxChoices', () => {
            expect(setChoicesSubset([], [{ id: '1' }, { id: '2' }], 0)).toEqual([]);
            expect(setChoicesSubset([], [{ id: '1' }, { id: '2' }], -1)).toEqual([]);
        });

        it('Should remove duplicates', () => {
            expect(setChoicesSubset([], [{ id: '1' }, { id: '2' }, { id: '1' }], Infinity)).toEqual([
                { id: '1' },
                { id: '2' },
            ]);
        });

        it('Should not modify object of nothing changed', () => {
            const choices = [{ id: '1' }, { id: '2' }];

            const subset = [{ id: '1' }, { id: '2' }];

            const actual = setChoicesSubset(choices, subset, Infinity);

            expect(actual).toEqual(subset);
            expect(actual).toBe(choices);
        });
    });

    describe('hasChoicesSubset()', () => {
        it('Should check if subset exists in choices', () => {
            expect(hasChoicesSubset([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '1' }, { id: '3' }])).toBe(true);
        });

        it('Should check if every choice from subset exists in choices', () => {
            expect(
                hasChoicesSubset([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '1' }, { id: '3' }, { id: '4' }]),
            ).toBe(false);
        });
    });

    describe('delChoicesSubset()', () => {
        it('Should remove subset', () => {
            expect(delChoicesSubset([{ id: '1' }, { id: '2' }], [{ id: '1' }], Infinity)).toEqual([{ id: '2' }]);
        });

        it('Should remove all', () => {
            expect(delChoicesSubset([{ id: '1' }, { id: '2' }], [{ id: '1' }, { id: '2' }], Infinity)).toEqual([]);
        });

        it('Should ignore non existing choices', () => {
            expect(delChoicesSubset([{ id: '1' }, { id: '2' }], [{ id: '3' }], Infinity)).toEqual([
                { id: '1' },
                { id: '2' },
            ]);
        });

        it('Should remove duplicates', () => {
            expect(delChoicesSubset([{ id: '1' }, { id: '1' }, { id: '2' }], [{ id: '1' }], Infinity)).toEqual([
                { id: '2' },
            ]);
        });

        it('Should always remove duplicates', () => {
            expect(delChoicesSubset([{ id: '1' }, { id: '1' }, { id: '2' }], [], Infinity)).toEqual([
                { id: '1' },
                { id: '2' },
            ]);
        });

        it('Should not modify choices if nothing changes', () => {
            const actual = delChoicesSubset([{ id: '1' }], [], Infinity);

            expect(actual).toEqual([{ id: '1' }]);

            expect(actual).toBe(actual);
        });

        it('Should also always crop choices to maxChoices from head', () => {
            expect(delChoicesSubset([{ id: '1' }, { id: '2' }, { id: '3' }], [], 1)).toEqual([{ id: '3' }]);
        });
    });

    describe('addChoicesSubset()', () => {
        it('Should add choices to subset', () => {
            expect(addChoicesSubset([], [{ id: '1' }, { id: '2' }], Infinity)).toEqual([{ id: '1' }, { id: '2' }]);
        });

        it('Should remove duplicates', () => {
            expect(addChoicesSubset([], [{ id: '1' }, { id: '2' }, { id: '1' }], Infinity)).toEqual([
                { id: '1' },
                { id: '2' },
            ]);
        });

        it('Should always remove duplicates', () => {
            expect(addChoicesSubset([{ id: '1' }, { id: '2' }, { id: '1' }], [], Infinity)).toEqual([
                { id: '1' },
                { id: '2' },
            ]);
        });

        it('Should crop choices to maxChoices from head', () => {
            expect(addChoicesSubset([], [{ id: '1' }, { id: '2' }, { id: '3' }], 1)).toEqual([{ id: '3' }]);
        });

        it('Should always crop choices to maxChoices from head', () => {
            expect(addChoicesSubset([{ id: '1' }, { id: '2' }, { id: '3' }], [], 1)).toEqual([{ id: '3' }]);
        });

        it('Should not modify choices if nothing changes', () => {
            const choices = [{ id: '1' }];

            const actual = addChoicesSubset(choices, [], Infinity);

            expect(actual).toEqual(choices);
            expect(actual).toBe(choices);
        });
    });

    describe('findPrevPicked()', () => {
        it('Should support undefined choices', () => {
            expect(findPrevPicked(undefined, [], false)).toEqual([]);
        });

        it('Should find prev picked choices', () => {
            expect(findPrevPicked([{ id: '1' }, { id: '2' }], [{ id: '2' }], false)).toEqual([{ id: '1' }]);
        });

        it('Should move from last picked choice', () => {
            expect(
                findPrevPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '1' }, { id: '2' }, { id: '3' }], false),
            ).toEqual([{ id: '2' }]);
        });

        it('Should support unsorted picked', () => {
            expect(
                findPrevPicked(
                    [{ id: '1' }, { id: '2' }, { id: '3' }, { id: '4' }],
                    [{ id: '4' }, { id: '2' }, { id: '3' }],
                    false,
                ),
            ).toEqual([{ id: '2' }]);
        });

        it('Should support top unsorted picked', () => {
            expect(
                findPrevPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '3' }, { id: '1' }, { id: '2' }], false),
            ).toEqual([{ id: '1' }]);
        });

        it('Should support top unsorted corner picked', () => {
            expect(
                findPrevPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '3' }, { id: '2' }, { id: '1' }], false),
            ).toEqual([]);
        });

        it('Should support empty picked', () => {
            expect(findPrevPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [], false)).toEqual([{ id: '3' }]);
        });

        it('Should support top corner', () => {
            expect(findPrevPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '1' }], false)).toEqual([]);
        });

        it('Should support appending', () => {
            expect(findPrevPicked([{ id: '1' }, { id: '2' }], [{ id: '2' }], true)).toEqual([{ id: '2' }, { id: '1' }]);
        });

        it('Should support top corner appending', () => {
            expect(findPrevPicked([{ id: '1' }, { id: '2' }], [{ id: '2' }, { id: '1' }], true)).toEqual([
                { id: '2' },
                { id: '1' },
            ]);
        });

        it('Should support range reducing', () => {
            expect(
                findPrevPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '1' }, { id: '2' }, { id: '3' }], true),
            ).toEqual([{ id: '1' }, { id: '2' }]);
        });

        it('Should support range filling', () => {
            expect(
                findPrevPicked(
                    [{ id: '1' }, { id: '2' }, { id: '3' }, { id: '4' }],
                    [{ id: '1' }, { id: '4' }, { id: '3' }],
                    true,
                ),
            ).toEqual([{ id: '1' }, { id: '4' }, { id: '3' }, { id: '2' }]);
        });

        it('Should support range skip filling', () => {
            expect(
                findPrevPicked(
                    [{ id: '1' }, { id: '2' }, { id: '3' }, { id: '4' }, { id: '5' }],
                    [{ id: '1' }, { id: '5' }, { id: '3' }, { id: '4' }],
                    true,
                ),
            ).toEqual([{ id: '1' }, { id: '5' }, { id: '3' }, { id: '4' }, { id: '2' }]);
        });
    });

    describe('findNextPicked()', () => {
        it('Should support undefined choices', () => {
            expect(findNextPicked(undefined, [], false)).toEqual([]);
        });

        it('Should find next picked choices', () => {
            expect(findNextPicked([{ id: '1' }, { id: '2' }], [{ id: '1' }], false)).toEqual([{ id: '2' }]);
        });

        it('Should move from last picked choice', () => {
            expect(
                findNextPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '3' }, { id: '2' }, { id: '1' }], false),
            ).toEqual([{ id: '2' }]);
        });

        it('Should support unsorted picked', () => {
            expect(
                findNextPicked(
                    [{ id: '1' }, { id: '2' }, { id: '3' }, { id: '4' }],
                    [{ id: '1' }, { id: '3' }, { id: '2' }],
                    false,
                ),
            ).toEqual([{ id: '3' }]);
        });

        it('Should support bottom unsorted picked', () => {
            expect(
                findNextPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '1' }, { id: '3' }, { id: '2' }], false),
            ).toEqual([{ id: '3' }]);
        });

        it('Should support bottom unsorted corner picked', () => {
            expect(
                findNextPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '1' }, { id: '2' }, { id: '3' }], false),
            ).toEqual([]);
        });

        it('Should support empty picked', () => {
            expect(findNextPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [], false)).toEqual([{ id: '1' }]);
        });

        it('Should support bottom corner', () => {
            expect(findNextPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '3' }], false)).toEqual([]);
        });

        it('Should support appending', () => {
            expect(findNextPicked([{ id: '1' }, { id: '2' }], [{ id: '1' }], true)).toEqual([{ id: '1' }, { id: '2' }]);
        });

        it('Should support bottom corner appending', () => {
            expect(findNextPicked([{ id: '1' }, { id: '2' }], [{ id: '1' }, { id: '2' }], true)).toEqual([
                { id: '1' },
                { id: '2' },
            ]);
        });

        it('Should support range reducing', () => {
            expect(
                findNextPicked([{ id: '1' }, { id: '2' }, { id: '3' }], [{ id: '3' }, { id: '2' }, { id: '1' }], true),
            ).toEqual([{ id: '3' }, { id: '2' }]);
        });

        it('Should support range filling', () => {
            expect(
                findNextPicked(
                    [{ id: '1' }, { id: '2' }, { id: '3' }, { id: '4' }],
                    [{ id: '4' }, { id: '1' }, { id: '2' }],
                    true,
                ),
            ).toEqual([{ id: '4' }, { id: '1' }, { id: '2' }, { id: '3' }]);
        });

        it('Should support range skip filling', () => {
            expect(
                findNextPicked(
                    [{ id: '1' }, { id: '2' }, { id: '3' }, { id: '4' }, { id: '5' }],
                    [{ id: '5' }, { id: '1' }, { id: '3' }, { id: '2' }],
                    true,
                ),
            ).toEqual([{ id: '5' }, { id: '1' }, { id: '3' }, { id: '2' }, { id: '4' }]);
        });
    });
});
