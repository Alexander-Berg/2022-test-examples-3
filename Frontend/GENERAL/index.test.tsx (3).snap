// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IotValueWithUnit Должно корректно отрендериться безразмерная величина 1`] = `
<Fragment>
  123.456
</Fragment>
`;

exports[`IotValueWithUnit Должно корректно отрендериться значение в амперах 1`] = `
<Fragment>
  0.32
  <IotUnit
    unit="unit.ampere"
  />
</Fragment>
`;

exports[`IotValueWithUnit Должно корректно отрендериться значение в ваттах 1`] = `
<Fragment>
  1344.82
  <IotUnit
    unit="unit.watt"
  />
</Fragment>
`;

exports[`IotValueWithUnit Должно корректно отрендериться значение в вольтах 1`] = `
<Fragment>
  223.1
  <IotUnit
    unit="unit.volt"
  />
</Fragment>
`;

exports[`IotValueWithUnit Должно корректно отрендериться значение в градусах Цельсия 1`] = `
<Fragment>
  238.333
  <IotUnit
    unit="unit.temperature.celsius"
  />
</Fragment>
`;

exports[`IotValueWithUnit Должно корректно отрендериться значение в кельвинах 1`] = `
<Fragment>
  291.111
  <IotUnit
    unit="unit.temperature.kelvin"
  />
</Fragment>
`;

exports[`IotValueWithUnit Должно корректно отрендериться значение в процентах 1`] = `
<Fragment>
  89
  <IotUnit
    unit="unit.percent"
  />
</Fragment>
`;

exports[`IotValueWithUnit Должно корректно отрендериться частиц на миллион 1`] = `
<Fragment>
  2336
  <IotUnit
    unit="unit.ppm"
  />
</Fragment>
`;
